### PR TITLE
[file-format] establish a common ground for visibility extensions

### PR DESCRIPTION
**Pull request by mengxu-fb**
## Motivation

This is an extraction of #7343 by removing all "friend" / protected stuff but retaining the changes to the bytecode format. The goal is to establish a common ground for future extension of visibility, being it "friend" or "script" or "package", etc.

With this PR:

- Visibility is not a boolean flag `is_public` anymore. It is an enum now, although there are only `Public` and `Private` in the enum.
- For serialization and deserialization, the visibility info is not mixed into a one-byte flag. Instead, the visibility info is stored a separate byte with 0x0 being 'private' and being 0x1 being `public`. The old flag byte now has only 0x2 bit in use, which is to store the `is_native` information. The bytecode version is updated to version 2 to reflect this change.

## Test Plan

All tests should pass, several tests are updated:

- `deserializer_tests.rs` are updated such that the `simple_malformed` test is applied to not only the most recent version, but also any prior versions.
- E2E tests that stores compiled module bytes in the golden files are updated (golden files regenerated)

### CODE REVIEW

File: language/vm/src/deserializer.rs
Comment (tnowacki) [inline language/vm/src/deserializer.rs:1123]: return Err(PartialVMError::new(StatusCode::UNREACHABLE))

Best not to accidentally crash the VM :P

File: language/vm/src/deserializer.rs
Comment (tnowacki) [inline language/vm/src/deserializer.rs:1094]: Something that isn't done in the old version either, but we should check for bits that are set that make no sense. I think it's definitely a bug in the current version, since if add meaning to the bits, suddenly a bunch of modules on chain might break.

Comment (mengxu-fb) [reply language/vm/src/deserializer.rs:1094]: Will add the check

Comment (mengxu-fb) [reply language/vm/src/deserializer.rs]: Check added in f4a3f50.

File: language/vm/src/deserializer.rs
Comment (tnowacki) [inline language/vm/src/deserializer.rs:1102]: Clever!

File: language/move-lang/src/interface_generator.rs
Comment (tnowacki) [inline language/move-lang/src/interface_generator.rs:63]: Fully pattern match, it will make migrating this easier as we add visibilities (or break out into a helper for something like `is_externally_visible` )

File: language/bytecode-verifier/src/dependencies.rs
Comment (tnowacki) [inline language/bytecode-verifier/src/dependencies.rs:107]: Fully pattern match. It will make updating for new visibilities easier

File: language/tools/disassembler/src/disassembler.rs
Comment (tnowacki) [inline language/tools/disassembler/src/disassembler.rs:856]: fully pattern match or break out into a helper. It will make updating easier

File: language/vm/src/views.rs
Comment (tnowacki) [inline language/vm/src/views.rs:481]: Switch this view to `pub fn visibility(&self) -> Visibility`

**Review by tnowacki — CHANGES_REQUESTED** [review]

File: language/vm/src/views.rs
Comment (tnowacki) [reply language/vm/src/views.rs:481]: I imagine you will have more places to clean after touching this, thats why hit request changes :)

Comment (mengxu-fb) [reply language/vm/src/views.rs:481]: It will inch into the `move-model` land which I have less confidence in. I'll also request reviews from Wolfgang after the change.

File: language/move-model/src/model.rs
Comment (mengxu-fb) [inline language/move-model/src/model.rs:2354]: @wrwg would you mind checking whether this change is OK to you and also suggest how we should further propagating the visibility information in the `move-model`? For example, shall we change the function `pub fn is_public(&self) -> bool` into `pub fn visibility(&self) -> Visibility`?

Comment (wrwg) [reply language/move-model/src/model.rs:2354]: I suggest to move forward with keeping this API as is and just adjust the impl as you do. We (prover team) can then come back and change the API once we use this new info.

**Review by wrwg — APPROVED** [review]
Changes to move-model look good to me.

**Comment by mengxu-fb** [comment]
@tnowacki waiting for your final say on the PR :)

**Comment by sblackshear** [comment]
Should we go ahead and add the `Script` visibility mode here?

**Comment by tnowacki** [comment]
> Should we go ahead and add the Script visibility mode here?

A different PR :) But this opens the way for it

**Review by tnowacki — APPROVED** [review]

**Comment by mengxu-fb** [comment]
Rebase past #7443 for cluster test

**Comment by mengxu-fb** [comment]
Rebase past #7449 for unit test

**Comment by mengxu-fb** [comment]
Rebase past #7436, merge conflict resolved

File: language/vm/src/serializer.rs
Comment (tnowacki) [inline language/vm/src/serializer.rs:42]: Why don't we just serialize it to the version we deserialized it from?

Comment (mengxu-fb) [reply language/vm/src/serializer.rs:42]: You mean, the latest version?

**Review by tnowacki — CHANGES_REQUESTED** [review]
Blocking for discussion on serialization (just making it doesn't go in just yet, changes may or may not be fine :))

File: language/vm/src/serializer.rs
Comment (tnowacki) [reply language/vm/src/serializer.rs:42]: No, I mean whatever we deserialized this from. We are deserializing it from Version X, why are we not re-serializing it back as Version X? Why is the version hardcoded?

**Comment by tnowacki** [comment]
I'm not sure what is going on with this versioned serializer. It is definitely not something we can support for abilities

File: language/vm/src/serializer.rs
Comment (mengxu-fb) [reply language/vm/src/serializer.rs:42]: I guess partially the reason is, we did not record the version number in `CompiledModule` :(

Comment (vgao1996) [reply language/vm/src/serializer.rs:42]: Given `CompiledScriptMut`, there doesn't necessarily exist a version X blob it was serialized to, right? I mean the script could have been created entirely in memory, for example, by the compiler.

Comment (mengxu-fb) [reply language/vm/src/serializer.rs:42]: Right, so we have to choose a version to serialize to.

**Comment by mengxu-fb** [comment]
Versioned serializer is dropped by the two "reverts", seems to be an overkill and can soon become intractable.

The serializer is still programmed to serialize in old format (in commit 6f064ad), just to make cluster test happy. This behavior will be reverted (by a follow-up PR) as soon as this PR is merged in.

**Comment by mengxu-fb** [comment]
Bring the PR to the state where @tnowacki last reviewed. Pending rebase after #7493 lands.

**Review by tnowacki — APPROVED** [review]

**Pull request closed without merging**
